plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+
+    while (read_nbytes > 0) {
+      TSIOBufferBlock blk = TSIOBufferReaderStart(reader);
+      char *buf           = (char *)TSIOBufferBlockReadStart(blk, reader, &avail);
+      int64_t read_ndone  = (avail >= read_nbytes) ? read_nbytes : avail;
+      int64_t consume     = read_ndone;
+      std::string chunk   = std::string(buf, read_ndone);
+
+      data->http_header += chunk;
+      // TSDebug(PLUGIN_NAME, "Headers: \n%s", icap_header.c_str());
+      int64_t pos          = data->http_header.find("\r\n\r\n");
+      int64_t token_length = std::string("\r\n").size();
+
+      if (pos != (int64_t)std::string::npos) {
+        data->http_header.resize(pos);
+        consume = pos + token_length - consumed;
+        TSIOBufferReaderConsume(reader, consume);
+        handle_icap_http_header(data);
+        return transform_read_http_body(contp, data);
+      }
+
+      if (read_ndone > 0) {
+        read_nbytes -= consume;
+        TSIOBufferReaderConsume(reader, consume);
+        consumed += consume;
+      } else
+        break;
+    }
+
+    if (read_nbytes <= 0) {
+      /* In case of finish reading http header, start reading http body length */
+      return transform_read_http_body(contp, data);
+    }
+  }
+  default:
+    break;
+  }
+
+  return 0;
+}
+
+static int
+transform_read_http_body_event(TSCont contp, TransformData *data, TSEvent event, void *edata)
+{
+  switch (event) {
+  case TS_EVENT_ERROR:
+    TSVConnAbort(data->icap_vc, 1);
+    data->icap_vc  = NULL;
+    data->icap_vio = NULL;
+
+    TSVConnAbort(data->output_vc, 1);
+    data->output_vc  = NULL;
+    data->output_vio = NULL;
+    break;
+  case TS_EVENT_VCONN_EOS:
+    TSVConnShutdown(data->icap_vc, 1, 0);
+    TSVIOReenable(data->output_vio);
+    data->eos_detected = true;
+    break;
+  case TS_EVENT_VCONN_READ_READY:
+    handle_read_http_body(contp, data);
+    TSVIOReenable(data->output_vio);
+    break;
+  case TS_EVENT_VCONN_WRITE_COMPLETE:
+    TSVConnShutdown(data->output_vc, 0, 1);
+    break;
+  case TS_EVENT_VCONN_WRITE_READY:
+    TSVIOReenable(data->icap_vio);
+    handle_read_http_body(contp, data);
+    break;
+  default:
+    break;
+  }
+
+  return 0;
+}
+
+static int
+transform_send_error_msg_event(TSCont contp, TransformData *data, TSEvent event, void *edata)
+{
+  switch (event) {
+  case TS_EVENT_VCONN_WRITE_COMPLETE:
+    TSVConnShutdown(data->output_vc, 0, 1);
+    break;
+  case TS_EVENT_VCONN_WRITE_READY:
+  default:
+    TSVIOReenable(data->output_vio);
+    break;
+  }
+
+  return 0;
+}
+
+static int
+transform_bypass_event(TSCont contp, TransformData *data, TSEvent event, void *edata)
+{
+  switch (event) {
+  case TS_EVENT_VCONN_WRITE_COMPLETE:
+    TSVConnShutdown(data->output_vc, 0, 1);
+    break;
+  default:
+    TSVIOReenable(data->output_vio);
+    break;
+  }
+
+  return 0;
+}
+
+/* Used only for debug-mode */
+static int
+transform_buffer_os_resp_event(TSCont contp, TransformData *data, TSEvent event, void *edata)
```

Comment:
I don't see that **event** or **edata** are used anywhere in this function.  Why are they included?  This clearly isn't a (direct) continuation function.  If **event** or **edata** were unused in a continuation function, I think it would be more readable if the parameter was left anonymous (making it clear without a doubt that it's unused).